File: tests/case_study/C++/speedtest_smooth_index_3d.cpp

#include <iomanip>
#include "far.h"
#include "timer.h"
#define NX 1000
#define NY 1000
#define NZ 1000
#define ITS 20

using namespace far;

int main(){
	Array<double,3> data(Range(0,NX+1),Range(0,NY+1),Range(0,NZ+1)), data2;
	data=0;
	data2.mold(data);
	//Set the boundaries on data
	data(0,Range(0,NY+1),Range(0,NZ+1))=1;
	data(NX+1,Range(0,NY+1),Range(0,NZ+1))=1;
	data(Range(0,NX+1),0,Range(0,NZ+1))=1;
	data(Range(0,NX+1),NY+1,Range(0,NZ+1))=1;
	data(Range(0,NX+1),Range(0,NY+1),0)=1;
	data(Range(0,NX+1),Range(0,NY+1),NZ+1)=1;


	timer t;
	t.begin("main loop");
#pragma omp parallel
	{
		for (int its = 0; its<ITS;++its){
			for (int k=1;k<=NZ;++k){
				for (int j=1;j<=NY;++j){
					#pragma omp for
					for (int i=1;i<=NX;++i){
						data2(i,j,k) = (
							data(i-1,j,k) + data(i+1,j,k) 
							+ data(i,j-1,k) + data(i,j+1,k)
							+ data(i,j,k-1) + data(i,j,k+1)
						) /6.0;
					}
				}
			}
			//Copy back
			data(Range(1,NX),Range(1,NY),Range(1,NZ))=data2(Range(1,NX),Range(1,NY),Range(1,NZ));
#pragma omp barrier
		}
	}
	t.end();
	std::cout << std::setprecision(17);
	std::cout << maxval(data(Range(1,NX),Range(1,NY),Range(1,NZ))) << "\n";
	std::cout << minval(data(Range(1,NX),Range(1,NY),Range(1,NZ))) << "\n";
	std::cout << sum(data(Range(1,NX),Range(1,NY),Range(1,NZ)))/float(data(Range(1,NX),Range(1,NY),Range(1,NZ)).getSize()) << "\n";
}

File: tests/case_study/Fortran/speedtest_smooth_index_3d.f90
PROGRAM jacobi
  USE ISO_FORTRAN_ENV
  IMPLICIT NONE
  INTEGER, PARAMETER :: NX=1000, NY=1000, NZ=1000, NITS=20
  REAL(REAL64), DIMENSION(:,:,:), ALLOCATABLE,TARGET :: data, data2
  INTEGER :: i, j, k, its, start,end
  REAL :: count_rate
  ALLOCATE(data(0:NX+1,0:NY+1,0:NZ+1), data2(0:NX+1,0:NY+1,0:NZ+1))
  data=0
  data(0,:,:)=1
  data(NX+1,:,:)=1
  data(:,0,:)=1
  data(:,NY+1,:)=1
  data(:,:,0)=1
  data(:,:,NZ+1)=1
  CALL SYSTEM_CLOCK(start,count_rate)
  DO its=1,NITS
    DO k=1,NZ
      DO j=1,NY
        DO i=1,NX
          data2(i,j,k)=(data(i-1,j,k)+data(i+1,j,k)+data(i,j-1,k)+data(i,j+1,k)+data(i,j,k-1)+data(i,j,k+1))/6.0_REAL64
        ENDDO
      ENDDO
    ENDDO
    data(1:NX,1:NY,1:NZ)=data2(1:NX,1:NY,1:NZ)
  ENDDO
  CALL SYSTEM_CLOCK(end)
  PRINT *, 'Time for main loop is ', REAL(end-start)/count_rate
  PRINT '(F24.17)', MAXVAL(data(1:NX,1:NY,1:NZ))
  PRINT '(F24.17)', MINVAL(data(1:NX,1:NY,1:NZ))
  PRINT '(F24.17)', SUM(data(1:NX,1:NY,1:NZ))/REAL(SIZE(data(1:NX,1:NY,1:NZ),KIND=INT64),REAL64)

END PROGRAM jacobi
